File: homework2/Saytam_Mishra/q7_copyTree.java
package homework2.Saytam_Mishra;
// Depth-first traversal (Pre-order)
// Time Complexity - O(n)
// Space Complexity - O(n)

public class q7_copyTree {
    static class TreeNode {
        TreeNode left;
        TreeNode right;
        int val;

        public TreeNode(int val) {
            this.left = null;
            this.right = null;
            this.val = val;
        }
    }

    public static TreeNode deepCopy(TreeNode root) {
        if(root == null) return null;

        TreeNode newNode = new TreeNode(root.val);

        newNode.left = deepCopy(root.left);
        newNode.right = deepCopy(root.right);

        return newNode;
    }

    public static void printInOrder(TreeNode node) {
        if(node == null) return;

        printInOrder(node.left);
        System.out.print(node.val + " ");
        printInOrder(node.right);
    }

    public static void main(String[] args) {
        // Test Case - 1: Change a value in the NEW tree
        
        TreeNode root1 = new TreeNode(10);
        root1.left = new TreeNode(15);
        root1.right = new TreeNode(20);

        System.out.print("Original tree values (inorder): ");
        printInOrder(root1);
        System.out.println();

        TreeNode copiedRoot1 = deepCopy(root1);
        
        copiedRoot1.left.val = 99; // Change 15 to 99

        System.out.println("\n--- After changing copied tree's left child to 99 ---");
        System.out.print("Original tree values (in-order): ");
        printInOrder(root1);
        System.out.println();
        System.out.print("Copied tree values (in-order):   ");
        printInOrder(copiedRoot1);

        // Test Case - 2: Copy empty Tree
        System.out.println("\n------");
        TreeNode root2 = null;
        TreeNode copiedRoot2 = deepCopy(root2);
        System.out.println("Original root: " + root2);
        System.out.println("Copied root:   " + copiedRoot2);
    }
}

// Time spent - 23m 30s

File: homework2/Aditya_Prasad/q7_CopyTree.cpp
/* Time Complexity : O(n) where n is the number of nodes in the tree 
   Memory Complexity : O(h) where h is the height of the tree (or it will be O(n) in case of skewed tree)
   Technique Used : Pre order Traversal to copy the nodes
   */

#include <bits/stdc++.h>
using namespace std;

struct Node{
    int data;
    Node* left;
    Node* right;

    Node(int val){
        data = val;
        left = nullptr;
        right = nullptr;
    }
};

Node* createCopy(Node* original) {
    if (original == NULL) return NULL;
    Node* newNode = new Node(original->data); 
    newNode->left = createCopy(original->left);
    newNode->right = createCopy(original->right);

    return newNode;
}



void inorder(Node* root){
    if(root == nullptr) return;
    inorder(root->left);
    cout << root->data << " ";
    inorder(root->right);
    return;
}

int main(){
    Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    root->right->left = new Node(6);
    root->right->right = new Node(7);

    

    inorder(root);

    cout << endl;

    Node* copy = createCopy(root);

    inorder(copy);
}

// Time : 10 mins